Gutenboarding: Persist empty form values

razvanpapadopol opened the pull request:
Value coming from persist is different than `EMPTY_FORM_VALUE` so it isn't considered 'empty'.

#### Changes proposed in this Pull Request

- Add `isEmpty` check to isFilledFormValue.

#### Testing instructions

**Before**:
![Screenshot 2019-11-20 at 12 47 47](https://user-images.githubusercontent.com/14192054/69232641-24e0d080-0b94-11ea-9721-4046c0ac9f2f.png)

**After**:
![Screenshot 2019-11-20 at 12 47 51](https://user-images.githubusercontent.com/14192054/69232655-28745780-0b94-11ea-8262-a5ebeea968c9.png)

#### Alternatives

- find a way to persist only values that are not empty

Commit:
Add isEmpty check to isFilledFormValue.
- value coming from persistance is different than EMPTY_FORM_VALUE

ockham commented:
> #### Alternatives
> * find a way to persist only values that are not empty

I'd prefer that TBH :grimacing:

razvanpapadopol commented:
> I'd prefer that TBH 😬

The problem is that we are providing the `persist` Array when calling `registerStore` so there is no way at the moment to persist something based on value of those state properties.

ockham commented:
> The problem is that we are providing the `persist` Array when calling `registerStore` so there is no way at the moment to persist something based on value of those state properties.

Right, makes sense.

However, I'd really rather not go with the PR's current solution (`isEmpty`) since IMO it obfuscates some underlying problem.

I think the invariant here should be that the initial value that we set in each individual reducer should also be what we use to persist. 

This might mean changing the definition of `EMPTY_FORM_VALUE`.

Let's pull in @sirreal

razvanpapadopol commented:
> the initial value that we set in each individual reducer should also be what we use to persist.
> This might mean changing the definition of `EMPTY_FORM_VALUE`.

Yes, I think we need something that can still be compared after serialisation/deserialisation that happens when saved and loaded from localStorage.

sirreal commented:
Interesting find, thanks @razvanpapadopol!

> > * find a way to persist only values that are not empty
> 
> I'd prefer that TBH 😬

I agree 👍 

> The problem is that we are providing the `persist` Array when calling `registerStore` so there is no way at the moment to persist something based on value of those state properties.

Yep, the persistence plugin doesn't seem to be very open to customizing how or if things are persisted, it may be a place we can provide an enhancement for Gutenberg, but I don't think we'll need it for this case…

> Yes, I think we need something that can still be compared after serialisation/deserialisation that happens when saved and loaded from localStorage.

Sounds about right. Did you have something in mind? The referential equality check that seemed appealing in the form value work is complicating this 🙂 

I was thinking something like:

```ts
EMPTY_FORM_VALUE = Object.freeze( { __EFV__: '__EFV__' } )
isFilledFormValue = v => v === EMPTY_FORM_VALUE || shallowEquals( v, EMPTY_FORM_VALUE )
```

Commit:
Add a shallow comparison to referential equality for isFilledFormValue.
- Assign a specific object to EMPTY_FORM_VALUE.
- use isShallowEqual to compare current value with EMPTY_FORM_VALUE.

razvanpapadopol commented on package.json:
```diff
@@ -67,6 +67,7 @@
 		"@wordpress/element": "2.9.0",
 		"@wordpress/format-library": "1.10.0",
 		"@wordpress/i18n": "3.7.0",
+		"@wordpress/is-shallow-equal": "1.6.1",
```
I see there is actively used `react-pure-render/shallowEqual` in Calypso but the package is no longer maintained and Gutenberg offers this alternative should be better.

sirreal replied:
Good spot: #37804

sirreal committed:
Add types for @wordpress/is-shallow-equal

sirreal reviewed (approved):
Nice catch and good fix 👍
